src/Chainweb/Pact/PactService/ExecBlock.hs

```diff
@@ -376,48 +380,42 @@ applyPactCmds
-    V.fromList . toList . sfst <$> V.foldM f (T2 mempty mc) cmds
-  where
-    f  (T2 dl mcache) cmd = applyPactCmd isGenesis env cmd miner mcache dl
-
--- | Apply a single Pact command
-applyPactCmd
-    :: Bool
-    -> P.PactDbEnv p
-    -> ChainwebTransaction
-    -> Miner
-    -> ModuleCache
-    -> DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))
-    -> PactServiceM cas
-       (T2 (DList (Either GasPurchaseFailure (P.CommandResult [P.TxLog A.Value]))) ModuleCache)
-applyPactCmd isGenesis dbEnv cmdIn miner mcache dl = do
-    logger <- view psLogger
-    gasModel <- view psGasModel
-    v <- view psVersion
-
-    handle onBuyGasFailure $ do
-      T2 result mcache' <- if isGenesis
-        then liftIO $! applyGenesisCmd logger dbEnv P.noSPVSupport (payloadObj <$> cmdIn)
-        else do
-          pd <- getTxContext (publicMetaOf $ payloadObj <$> cmdIn)
-          spv <- use psSpvSupport
-          liftIO $! applyCmd v logger dbEnv miner (gasModel pd) pd spv cmdIn mcache
-
-      when isGenesis $
-        updateInitCache mcache'
-
-      unless isGenesis $ debugResult "applyPactCmd" result
-
-      cp <- getCheckpointer
-      -- mark the tx as processed at the checkpointer.
-      liftIO $ _cpRegisterProcessedTx cp (P._cmdHash cmdIn)
-      pure $! T2 (DL.snoc dl (Right result)) mcache'
-
+applyPactCmds isGenesis env cmds miner mc blockGas = do
+    evalStateT (V.mapM go cmds) (T2 mc blockGas)
   where
-    onBuyGasFailure (BuyGasFailure f) = pure $! T2 (DL.snoc dl (Left f)) mcache
-    onBuyGasFailure e = throwM e
+    go cmd = StateT $ \(T2 mcache blockGasRemaining) -> do
+      logger <- view psLogger
+      gasModel <- view psGasModel
+      v <- view psVersion
+      let
+        onBuyGasFailure (BuyGasFailure f) = pure $! (Left f, T2 mcache blockGasRemaining)
+        onBuyGasFailure e = throwM e
+        gasLimitedCmd =
+          (payloadObj <$> cmd) & P.cmdPayload . P.pMeta . P.pmGasLimit %~ maybe id min (fromIntegral <$> blockGasRemaining)
+        initialGas = initialGasOf (P._cmdPayload cmd)
+      handle onBuyGasFailure $ do
+        T2 result mcache' <- if isGenesis
+          then liftIO $! applyGenesisCmd logger env P.noSPVSupport gasLimitedCmd
+          else do
+            pd <- getTxContext (publicMetaOf gasLimitedCmd)
+            spv <- use psSpvSupport
+            liftIO $! applyCmd v logger env miner (gasModel pd) pd spv gasLimitedCmd initialGas mcache
+
+        if isGenesis
+        then updateInitCache mcache'
+        else debugResult "applyPactCmd" result
+
+        cp <- getCheckpointer
+
+        -- mark the tx as processed at the checkpointer.
+        liftIO $ _cpRegisterProcessedTx cp (P._cmdHash cmd)
+        let blockGasRemaining' = (\g -> g - P._crGas result) <$> blockGasRemaining
+        case blockGasRemaining' of
+          Nothing -> return ()
+          Just r -> when (r <= 0) $ throwM $ BlockGasLimitExceeded (fromJuste blockGas - r)
```

Comment:
This is an issue. We use min(remaining block gas limit, tx gas limit) as the real tx gas limit to get early failure. This does its job, but if the tx consumes the entire gas limit, for example by failing, weird things happen. In particular, we have no way of distinguishing "block gas limit exceeded" from "tx failed". Also, the tx will use less than its real gas limit when it fails, which means the gas used will disagree with newBlock.

Reply:
Fixed now.